app/src/test/java/com/eveningoutpost/dexdrip/Models/BlueReaderTest.java

```diff
@@ -0,0 +1,556 @@
+        byte[] reply = blueReader.decodeblueReaderPacket(buffer, -1);
+
+        // :: Verify
+        assertThat(reply).isNull();
+    }
+
+    @Test
+    public void decodeBlueReaderPacket_IDR_noBgReading() {
+        // :: Setup
+        String inputString = "IDR0blue131-a1";
+
+        // :: Act
+        byte[] reply = blueReader.decodeblueReaderPacket(inputString.getBytes(), -1);
+
+        // :: Verify
+        assertThat(reply).isNotNull();
+        assertThat(new String(reply)).isEqualTo("l");
+        assertThat(PersistentStore.getString("blueReaderFirmware")).isEqualTo(inputString);
+    }
+
+    @Test
+    public void decodeBlueReaderPacket_IDR_OldBgReading() {
+        // :: Setup
+        String inputString = "IDR0blue131-a1";
+
+        // Add mock bg readings
+        Sensor mockSensor = createMockSensor();
+        addMockBgReading(125, 10, mockSensor);
+
+        // :: Act
+        byte[] reply = blueReader.decodeblueReaderPacket(inputString.getBytes(), -1);
+
+        // :: Verify
+        assertThat(reply).isNotNull();
+        assertThat(new String(reply)).isEqualTo("l");
+        assertThat(PersistentStore.getString("blueReaderFirmware")).isEqualTo(inputString);
+    }
+
+    @Test
+    public void decodeBlueReaderPacket_IDR_NewBgReading() {
+        // :: Setup
+        String inputString = "IDR0|blue131-a1";
+
+        // Add mock bg readings
+        Sensor mockSensor = createMockSensor();
+        addMockBgReading(125, 2, mockSensor);
+
+        // :: Act
+        byte[] reply = blueReader.decodeblueReaderPacket(inputString.getBytes(), -1);
+
+        // :: Verify
+        assertThat(reply).isNull();
+        assertThat(PersistentStore.getString("blueReaderFirmware")).isEqualTo(inputString);
+    }
+
+    @Test
+    public void decodeBlueReaderPacket_null() {
+        // :: Act
+        byte[] reply = blueReader.decodeblueReaderPacket(null, -1);
+
+        // :: Verify
+        assertThat(reply).isNull();
+        assertThat(getLogs()).contains("E/blueReader: null buffer passed to decodeblueReaderPacket");
+    }
+
+    // ===== Hibernate / reset tests ===============================================================
+
+    @Test
+    public void decodeBlueReaderPacket_notReadyFor_oneTime() {
+        // :: Setup
+        byte[] buffer = "not ready for".getBytes();
+
+        // :: Act
+        byte[] reply = blueReader.decodeblueReaderPacket(buffer, -1);
+
+        // :: Verify
+        assertThat(reply).isNotNull();
+        assertThat(new String(reply)).isEqualTo("h");
+        assertThat(getLogs()).contains("E/blueReader: Found blueReader in a ugly State (1/5), send hibernate to reset! If this does not help in the next 5 Minutes, then turn the bluereader manually off and on!");
+    }
+
+    /**
+     * Found a bug in the code which increase hibernate count twice pr. invocation instead of once.
+     * This test shows how it works. If This test fails because the bug has been fixed, rewrite the
+     * test.
```

Comment:
Thanks for the Hint. I seen that also in my last codereview, and will take care of that also for the Test within my next pull request. Also in future the count will raise only up to 3 as this is enough for checking the ugly state.

Reply:
Only wanted to write the test in this iteration, but it is easier to understand and remember when adding some documentation :-)